File: runtimes/java/spoi_accessor.java
// ============================================================
// SPOI Accessor — Java 类型特化访问器（自动生成）
// 由 sp-gen spoi-java-accessor 从 C++ 元数据生成
// 替代反射机制，直接通过字段索引访问/设置值
// ============================================================

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.*;

// ============================================================
// 基本类型 ID（与 C++ E_type 枚举值一致）
// ============================================================

class TypeId {
    public static final long U8     = 26L;
    public static final long U16    = 27L;
    public static final long U32    = 28L;
    public static final long U64    = 29L;
    public static final long I8     = 30L;
    public static final long I16    = 31L;
    public static final long I32    = 32L;
    public static final long I64    = 33L;
    public static final long F32    = 34L;
    public static final long F64    = 35L;
    public static final long STRING = 9L;
    public static final long BOOL   = 40L;
}

// ============================================================
// SpoiAccessor — 类型特化访问器接口
// ============================================================

interface SpoiAccessor {
    int fieldCount();
    Object getField(Object obj, int idx);
    void setField(Object obj, int idx, Object val);
}

// ============================================================
// DeserializeValue — 通用值反序列化（基于 type_id 前缀）
// 格式: [type_id(u32 LE) + value_bytes]
// ============================================================

class SpoiDeserializer {
    static Object deserializeValue(byte[] data) {
        if (data == null || data.length < 4) return null;
        long typeId = Integer.toUnsignedLong(
            ByteBuffer.wrap(data, 0, 4).order(ByteOrder.LITTLE_ENDIAN).getInt());
        byte[] valueBytes = Arrays.copyOfRange(data, 4, data.length);
        if (typeId == TypeId.U8)   return valueBytes[0] & 0xFF;
        if (typeId == TypeId.U16)  return (int)(ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getShort() & 0xFFFF);
        if (typeId == TypeId.U32)  return Integer.toUnsignedLong(ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getInt());
        if (typeId == TypeId.U64)  return ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getLong();
        if (typeId == TypeId.I8)   return valueBytes[0];
        if (typeId == TypeId.I16)  return ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getShort();
        if (typeId == TypeId.I32)  return ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getInt();
        if (typeId == TypeId.I64)  return ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getLong();
        if (typeId == TypeId.F32)  return ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getFloat();
        if (typeId == TypeId.F64)  return ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).getDouble();
        if (typeId == TypeId.STRING) return new String(valueBytes, StandardCharsets.UTF_8);
        if (typeId == TypeId.BOOL) return valueBytes[0] != 0;
        return valueBytes;
    }
}

// ============================================================
// SpoiTestPlayerAccessor
// ============================================================

class SpoiTestPlayerAccessor implements SpoiAccessor {

    public int fieldCount() { return 4; }

    public Object getField(Object obj, int idx) {
        SpoiTestPlayer o = (SpoiTestPlayer) obj;
        switch (idx) {
            case 0: return o.name;
            case 1: return Integer.valueOf(o.hp);
            case 2: return Integer.valueOf(o.level);
            case 3: return Double.valueOf(o.posX);
            default: throw new IllegalArgumentException("invalid field index for SpoiTestPlayer: " + idx);
        }
    }

    public void setField(Object obj, int idx, Object val) {
        SpoiTestPlayer o = (SpoiTestPlayer) obj;
        switch (idx) {
            case 0: o.name = val != null ? val.toString() : ""; break;
            case 1: o.hp = val instanceof Number ? ((Number)val).intValue() : 0; break;
            case 2: o.level = val instanceof Number ? ((Number)val).intValue() : 0; break;
            case 3: o.posX = val instanceof Number ? ((Number)val).doubleValue() : 0.0; break;
            default: throw new IllegalArgumentException("invalid field index for SpoiTestPlayer: " + idx);
        }
    }
}

// ============================================================
// SpoiTestStateAccessor
// ============================================================

class SpoiTestStateAccessor implements SpoiAccessor {

    public int fieldCount() { return 3; }

    public Object getField(Object obj, int idx) {
        SpoiTestState o = (SpoiTestState) obj;
        switch (idx) {
            case 0: return Integer.valueOf(o.tick);
            case 1: return o.currentMap;
            case 2: return o.players;
            default: throw new IllegalArgumentException("invalid field index for SpoiTestState: " + idx);
        }
    }

    public void setField(Object obj, int idx, Object val) {
        SpoiTestState o = (SpoiTestState) obj;
        switch (idx) {
            case 0: o.tick = val instanceof Number ? ((Number)val).intValue() : 0; break;
            case 1: o.currentMap = val != null ? val.toString() : ""; break;
            case 2: o.players = val; break;
            default: throw new IllegalArgumentException("invalid field index for SpoiTestState: " + idx);
        }
    }
}

// ============================================================
// SpoiItemAccessor
// ============================================================

class SpoiItemAccessor implements SpoiAccessor {

    public int fieldCount() { return 2; }

    public Object getField(Object obj, int idx) {
        SpoiItem o = (SpoiItem) obj;
        switch (idx) {
            case 0: return o.name;
            case 1: return Integer.valueOf(o.value);
            default: throw new IllegalArgumentException("invalid field index for SpoiItem: " + idx);
        }
    }

    public void setField(Object obj, int idx, Object val) {
        SpoiItem o = (SpoiItem) obj;
        switch (idx) {
            case 0: o.name = val != null ? val.toString() : ""; break;
            case 1: o.value = val instanceof Number ? ((Number)val).intValue() : 0; break;
            default: throw new IllegalArgumentException("invalid field index for SpoiItem: " + idx);
        }
    }
}

// ============================================================
// SpoiInventoryAccessor
// ============================================================

class SpoiInventoryAccessor implements SpoiAccessor {

    public int fieldCount() { return 3; }

    public Object getField(Object obj, int idx) {
        SpoiInventory o = (SpoiInventory) obj;
        switch (idx) {
            case 0: return o.items;
            case 1: return o.equipped;
            case 2: return Integer.valueOf(o.gold);
            default: throw new IllegalArgumentException("invalid field index for SpoiInventory: " + idx);
        }
    }

    public void setField(Object obj, int idx, Object val) {
        SpoiInventory o = (SpoiInventory) obj;
        switch (idx) {
            case 0: o.items = val; break;
            case 1: o.equipped = val; break;
            case 2: o.gold = val instanceof Number ? ((Number)val).intValue() : 0; break;
            default: throw new IllegalArgumentException("invalid field index for SpoiInventory: " + idx);
        }
    }
}

// ============================================================
// SpoiCharacterAccessor
// ============================================================

class SpoiCharacterAccessor implements SpoiAccessor {

    public int fieldCount() { return 5; }

    public Object getField(Object obj, int idx) {
        SpoiCharacter o = (SpoiCharacter) obj;
        switch (idx) {
            case 0: return o.name;
            case 1: return Integer.valueOf(o.hp);
            case 2: return o.inventory;
            case 3: return o.weapon;
            case 4: return Integer.valueOf(o.petLevel);
            default: throw new IllegalArgumentException("invalid field index for SpoiCharacter: " + idx);
        }
    }

    public void setField(Object obj, int idx, Object val) {
        SpoiCharacter o = (SpoiCharacter) obj;
        switch (idx) {
            case 0: o.name = val != null ? val.toString() : ""; break;
            case 1: o.hp = val instanceof Number ? ((Number)val).intValue() : 0; break;
            case 2: o.inventory = val; break;
            case 3: o.weapon = val; break;
            case 4: o.petLevel = val instanceof Number ? ((Number)val).intValue() : 0; break;
            default: throw new IllegalArgumentException("invalid field index for SpoiCharacter: " + idx);
        }
    }
}

// ============================================================
// SpoiWorldAccessor
// ============================================================

class SpoiWorldAccessor implements SpoiAccessor {

    public int fieldCount() { return 3; }

    public Object getField(Object obj, int idx) {
        SpoiWorld o = (SpoiWorld) obj;
        switch (idx) {
            case 0: return o.worldName;
            case 1: return Integer.valueOf(o.tick);
            case 2: return o.characters;
            default: throw new IllegalArgumentException("invalid field index for SpoiWorld: " + idx);
        }
    }

    public void setField(Object obj, int idx, Object val) {
        SpoiWorld o = (SpoiWorld) obj;
        switch (idx) {
            case 0: o.worldName = val != null ? val.toString() : ""; break;
            case 1: o.tick = val instanceof Number ? ((Number)val).intValue() : 0; break;
            case 2: o.characters = val; break;
            default: throw new IllegalArgumentException("invalid field index for SpoiWorld: " + idx);
        }
    }
}

// ============================================================
// SpoiAccessorRegistry — 静态类型注册表
// 替代运行时 Map<String, List<String>>（反射版本）
// ============================================================

class SpoiAccessorRegistry {
    static final Map<String, SpoiAccessor> registry = new HashMap<>();
    static {
        registry.put("SpoiTestPlayer", new SpoiTestPlayerAccessor());
        registry.put("SpoiTestState", new SpoiTestStateAccessor());
        registry.put("SpoiItem", new SpoiItemAccessor());
        registry.put("SpoiInventory", new SpoiInventoryAccessor());
        registry.put("SpoiCharacter", new SpoiCharacterAccessor());
        registry.put("SpoiWorld", new SpoiWorldAccessor());
    }
    static SpoiAccessor get(String typeName) { return registry.get(typeName); }
}

File: runtimes/kotlin/spoi_accessor.kt
// ============================================================
// SPOI Accessor — Kotlin 类型特化访问器（自动生成）
// 由 sp-gen spoi-kotlin-accessor 从 C++ 元数据生成
// 替代反射机制，直接通过字段索引访问/设置值
// ============================================================

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

// ============================================================
// 基本类型 ID（与 C++ E_type 枚举值一致）
// ============================================================

object TypeId {
    const val U8: Long     = 26L
    const val U16: Long    = 27L
    const val U32: Long    = 28L
    const val U64: Long    = 29L
    const val I8: Long     = 30L
    const val I16: Long    = 31L
    const val I32: Long    = 32L
    const val I64: Long    = 33L
    const val F32: Long    = 34L
    const val F64: Long    = 35L
    const val STRING: Long = 9L
    const val BOOL: Long   = 40L
}

// ============================================================
// SpoiAccessor — 类型特化访问器接口
// ============================================================

interface SpoiAccessor {
    fun fieldCount(): Int
    fun getField(obj: Any, idx: Int): Any?
    fun setField(obj: Any, idx: Int, value: Any?)
}

// ============================================================
// DeserializeValue — 通用值反序列化（基于 type_id 前缀）
// 格式: [type_id(u32 LE) + value_bytes]
// ============================================================

object SpoiDeserializer {
    fun deserializeValue(data: ByteArray): Any? {
        if (data.size < 4) return null
        val typeId = ByteBuffer.wrap(data, 0, 4).order(ByteOrder.LITTLE_ENDIAN).int.toLong() and 0xFFFFFFFFL
        val valueBytes = data.copyOfRange(4, data.size)
        return when (typeId) {
            TypeId.U8     -> valueBytes[0].toInt() and 0xFF
            TypeId.U16    -> (ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).short.toInt() and 0xFFFF)
            TypeId.U32    -> (ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).int.toLong() and 0xFFFFFFFFL)
            TypeId.U64    -> ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).long
            TypeId.I8     -> valueBytes[0]
            TypeId.I16    -> ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).short
            TypeId.I32    -> ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).int
            TypeId.I64    -> ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).long
            TypeId.F32    -> ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).float
            TypeId.F64    -> ByteBuffer.wrap(valueBytes).order(ByteOrder.LITTLE_ENDIAN).double
            TypeId.STRING -> String(valueBytes, StandardCharsets.UTF_8)
            TypeId.BOOL   -> valueBytes[0].toInt() != 0
            else          -> valueBytes
        }
    }
}

// ============================================================
// SpoiTestPlayerAccessor
// ============================================================

class SpoiTestPlayerAccessor : SpoiAccessor {

    override fun fieldCount(): Int = 4

    override fun getField(obj: Any, idx: Int): Any? {
        val o = obj as SpoiTestPlayer
        return when (idx) {
            0 -> o.name
            1 -> o.hp
            2 -> o.level
            3 -> o.posX
            else -> throw IllegalArgumentException("invalid field index for SpoiTestPlayer: $idx")
        }
    }

    override fun setField(obj: Any, idx: Int, value: Any?) {
        val o = obj as SpoiTestPlayer
        when (idx) {
            0 -> o.name = value?.toString() ?: ""
            1 -> o.hp = (value as? Number)?.toInt() ?: 0
            2 -> o.level = (value as? Number)?.toInt() ?: 0
            3 -> o.posX = (value as? Number)?.toDouble() ?: 0.0
            else -> throw IllegalArgumentException("invalid field index for SpoiTestPlayer: $idx")
        }
    }
}

// ============================================================
// SpoiTestStateAccessor
// ============================================================

class SpoiTestStateAccessor : SpoiAccessor {

    override fun fieldCount(): Int = 3

    override fun getField(obj: Any, idx: Int): Any? {
        val o = obj as SpoiTestState
        return when (idx) {
            0 -> o.tick
            1 -> o.currentMap
            2 -> o.players
            else -> throw IllegalArgumentException("invalid field index for SpoiTestState: $idx")
        }
    }

    override fun setField(obj: Any, idx: Int, value: Any?) {
        val o = obj as SpoiTestState
        when (idx) {
            0 -> o.tick = (value as? Number)?.toInt() ?: 0
            1 -> o.currentMap = value?.toString() ?: ""
            2 -> o.players = value
            else -> throw IllegalArgumentException("invalid field index for SpoiTestState: $idx")
        }
    }
}

// ============================================================
// SpoiItemAccessor
// ============================================================

class SpoiItemAccessor : SpoiAccessor {

    override fun fieldCount(): Int = 2

    override fun getField(obj: Any, idx: Int): Any? {
        val o = obj as SpoiItem
        return when (idx) {
            0 -> o.name
            1 -> o.value
            else -> throw IllegalArgumentException("invalid field index for SpoiItem: $idx")
        }
    }

    override fun setField(obj: Any, idx: Int, value: Any?) {
        val o = obj as SpoiItem
        when (idx) {
            0 -> o.name = value?.toString() ?: ""
            1 -> o.value = (value as? Number)?.toInt() ?: 0
            else -> throw IllegalArgumentException("invalid field index for SpoiItem: $idx")
        }
    }
}

// ============================================================
// SpoiInventoryAccessor
// ============================================================

class SpoiInventoryAccessor : SpoiAccessor {

    override fun fieldCount(): Int = 3

    override fun getField(obj: Any, idx: Int): Any? {
        val o = obj as SpoiInventory
        return when (idx) {
            0 -> o.items
            1 -> o.equipped
            2 -> o.gold
            else -> throw IllegalArgumentException("invalid field index for SpoiInventory: $idx")
        }
    }

    override fun setField(obj: Any, idx: Int, value: Any?) {
        val o = obj as SpoiInventory
        when (idx) {
            0 -> o.items = value
            1 -> o.equipped = value
            2 -> o.gold = (value as? Number)?.toInt() ?: 0
            else -> throw IllegalArgumentException("invalid field index for SpoiInventory: $idx")
        }
    }
}

// ============================================================
// SpoiCharacterAccessor
// ============================================================

class SpoiCharacterAccessor : SpoiAccessor {

    override fun fieldCount(): Int = 5

    override fun getField(obj: Any, idx: Int): Any? {
        val o = obj as SpoiCharacter
        return when (idx) {
            0 -> o.name
            1 -> o.hp
            2 -> o.inventory
            3 -> o.weapon
            4 -> o.petLevel
            else -> throw IllegalArgumentException("invalid field index for SpoiCharacter: $idx")
        }
    }

    override fun setField(obj: Any, idx: Int, value: Any?) {
        val o = obj as SpoiCharacter
        when (idx) {
            0 -> o.name = value?.toString() ?: ""
            1 -> o.hp = (value as? Number)?.toInt() ?: 0
            2 -> o.inventory = value
            3 -> o.weapon = value
            4 -> o.petLevel = (value as? Number)?.toInt() ?: 0
            else -> throw IllegalArgumentException("invalid field index for SpoiCharacter: $idx")
        }
    }
}

// ============================================================
// SpoiWorldAccessor
// ============================================================

class SpoiWorldAccessor : SpoiAccessor {

    override fun fieldCount(): Int = 3

    override fun getField(obj: Any, idx: Int): Any? {
        val o = obj as SpoiWorld
        return when (idx) {
            0 -> o.worldName
            1 -> o.tick
            2 -> o.characters
            else -> throw IllegalArgumentException("invalid field index for SpoiWorld: $idx")
        }
    }

    override fun setField(obj: Any, idx: Int, value: Any?) {
        val o = obj as SpoiWorld
        when (idx) {
            0 -> o.worldName = value?.toString() ?: ""
            1 -> o.tick = (value as? Number)?.toInt() ?: 0
            2 -> o.characters = value
            else -> throw IllegalArgumentException("invalid field index for SpoiWorld: $idx")
        }
    }
}

// ============================================================
// SpoiAccessorRegistry — 静态类型注册表
// 替代运行时 Map<String, List<String>>（反射版本）
// ============================================================

object SpoiAccessorRegistry {
    val registry: Map<String, SpoiAccessor> = mapOf(
        "SpoiTestPlayer" to SpoiTestPlayerAccessor(),
        "SpoiTestState" to SpoiTestStateAccessor(),
        "SpoiItem" to SpoiItemAccessor(),
        "SpoiInventory" to SpoiInventoryAccessor(),
        "SpoiCharacter" to SpoiCharacterAccessor(),
        "SpoiWorld" to SpoiWorldAccessor()
    )
    fun get(typeName: String): SpoiAccessor? = registry[typeName]
}
